#272 Handle errors that terminate state machines

aburkleaux-amazon opened the pull request:
## Issue #, if available: 
#272 

## Description of changes:

Added a new lambda resource, `WorkflowErrorHandlerLambda`, that is triggered by  `Step Functions Execution Status Change` EventBridge events that have an error status (`FAILED, TIMED_OUT, ABORTED`).

The `WorkflowErrorHandlerLambda` extracts the error information from the state machine execution history and propagates the error to the MIE control plane.

In order to handle multiple MIE stacks deployed in the same region, the event handler checks for the stack SHORT_UUID in the workflow name.  If the workflow name does not contain the stack, then it returns without handling the error.  The alternative to this would be to use the "prefix" event filter on the workflow name.  In order to do this, we would need to change the naming of workflows to be SHORT_UUID+WorkflowName instead of WorkflowName+SHORT_UUID.

I have a separate pull request to update the implementation guide with error handling info.
See also: https://github.com/awslabs/aws-media-insights/pull/24

## Testing:

I tested the error handler e2e manually with ABORTED and TIME_OUT events.  I simulated a failure event by pasting in event JSON to the lambda function.

## Future work:

#271 Add a configurable workflow timeout to the create workflow API
#270 Refactor workflowapi and workflow lambdas to use a library import for shared code 
#269 Automated integration test for workflow state machine execution failure that immediately terminate the execution 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

brandold commented on source/workflow/app.py:
```diff
@@ -52,6 +52,11 @@
 
+if "ShortUUID" in os.environ:
+    ShortUUID = os.environ["ShortUUID"]
+else:
```
Might be clearer to remove this else statement if it isn't doing anything. Not sure what the quotes are for.

brandold commented on source/workflow/app.py:
```diff
@@ -542,4 +547,109 @@ def update_workflow_execution_status(id, status, message):
+    
+    logger.info("workflow_error_handler_lambda: {}".format(json.dumps(event))) 
+    
+    if ShortUUID == "":
```
Related to comment on line 57. This could be simplified with:

`if os.environ["ShortUUID"]:
    logger.info("Process step function error event for stack with ShortUUID {}".format(os.environ["ShortUUID"]))
else:
   raise Exception('ShortUUID is not set in lambda environment.') 
`

This keeps us from having to do any string validation or store extra variables since it will return a boolean by default.

brandold reviewed (dismissed):
Overall this looks good to me. My only concerns are with the code style regarding the stack short UUID. I put some single comments here and there but more generally I'm curious if there is even a case where the UUID wouldn't be available.

aburkleaux-amazon commented:
> Overall this looks good to me. My only concerns are with the code style regarding the stack short UUID. I put some single comments here and there but more generally I'm curious if there is even a case where the UUID wouldn't be available.

When I added that code I was thinking about the fact that the code in the file supports a number of lambda functions and not all of them have the same environment variables set.  I added #270 (Refactor workflowapi and workflow lambdas to use a library import for shared code) to address the code organization issue in the future.  None of the other lambdas are calling this code, so it is pretty safe to assume the SHORT_UUID is set.  I'll make the change to simplify this new code.

aburkleaux-amazon commented:
Updated the pull request to address review comments from Brandon.
